File: Week_06/G20200343040044/minimum-path-sum.java
package 第五周;

class Solution {
    public int minPathSum(int[][] grid) {
     int[][] memo = new int[grid.length][grid[0].length];
     return minPathSumHelper(grid, 0, 0, memo);
    }
 
    public int minPathSumHelper(int[][] grid, int row, int col, int[][] memo) {
     if(row == grid.length-1 && col == grid[0].length-1) return grid[row][col];
     if(memo[row][col] != 0) return memo[row][col];
 
     int rowInc = Integer.MAX_VALUE, colInc = Integer.MAX_VALUE;
     if(row < grid.length-1) rowInc = minPathSumHelper(grid, row+1, col, memo);
     if(col < grid[0].length-1) colInc = minPathSumHelper(grid, row, col+1, memo);
     memo[row][col] = Math.min(rowInc, colInc) + grid[row][col];
     return memo[row][col];
   }
 }

File: Week_06/G20200343040030/minimum-path-sum.js
/**
 * dp方程：dp[i][j] = min(dp[i][j-1], dp[i-1][j]) + dp[i][j];
 * 边界条件处理：i=0或者j=0的时候
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function (grid) {
  let m = grid.length;
  let n = grid[0].length;
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (i == 0 && j == 0) continue;
      else if (i === 0) grid[i][j] = grid[i][j - 1] + grid[i][j];
      else if (j === 0) grid[i][j] = grid[i - 1][j] + grid[i][j];
      else grid[i][j] = Math.min(grid[i - 1][j], grid[i][j - 1]) + grid[i][j];
    }
  }
  return grid[m - 1][n - 1];
};
